prow/config/cache.go

```diff
@@ -0,0 +1,219 @@
+
+// ProwYAMLCache is the user-facing cache. It acts as a wrapper around the
+// generic LRUCache, by handling type casting in and out of the LRUCache (which
+// only handles empty interfaces).
+type ProwYAMLCache cache.LRUCache
+
+// NewProwYAMLCache creates a new LRU cache for ProwYAML values, where the keys
+// are CacheKeys and values are pointers to ProwYAMLs.
+func NewProwYAMLCache(size int) (*ProwYAMLCache, error) {
+	cache, err := cache.NewLRUCache(size)
+	if err != nil {
+		return nil, err
+	}
+
+	return (*ProwYAMLCache)(cache), nil
+}
+
+// CacheKey acts as a key to the ProwYAMLCache. We construct it by marshaling
+// CacheKeyParts into a JSON string.
+type CacheKey string
+
+// The CacheKeyParts is a struct because we want to keep the various components
+// that make up the key separate to help keep tests readable. Because the
+// headSHAs field is a slice, the overall CacheKey object is not hashable and
+// cannot be used directly as a key. Instead we marshal it to JSON first, then
+// convert its type to CacheKey.
+//
+// Users should take care to ensure that headSHAs remains stable (order
+// matters).
+type CacheKeyParts struct {
+	Identifier string   `json:"identifier"`
+	BaseSHA    string   `json:"baseSHA"`
+	HeadSHAs   []string `json:"headSHAs"`
+}
+
+// MakeCacheKeyParts constructs a CacheKeyParts struct from uniquely-identifying
+// information.
+func MakeCacheKeyParts(
+	identifier string,
+	baseSHAGetter RefGetter,
+	headSHAGetters ...RefGetter) (CacheKeyParts, error) {
+	// Initialize empty key parts.
+	keyParts := CacheKeyParts{}
+
+	// Append "identifier" string information.
+	if identifier == "" {
+		return CacheKeyParts{}, fmt.Errorf("no identifier for repo given")
+	}
+	keyParts.Identifier = identifier
+
+	// Append "baseSHA" string information.
+	baseSHA, err := baseSHAGetter()
+	if err != nil {
+		return CacheKeyParts{}, fmt.Errorf("failed to get baseSHA: %v", err)
+	}
+	keyParts.BaseSHA = baseSHA
+
+	// Append "headSHAs" string information.
+	var headSHAs []string
+	for _, headSHAGetter := range headSHAGetters {
+		headSHA, err := headSHAGetter()
+		if err != nil {
+			return CacheKeyParts{}, fmt.Errorf("failed to get headRef: %v", err)
+		}
+		headSHAs = append(headSHAs, headSHA)
+	}
+
+	keyParts.HeadSHAs = headSHAs
+
+	return keyParts, nil
+}
+
+// MakeCacheKey converts a CacheKeyParts object into a JSON string (to be used
+// as a CacheKey).
+func MakeCacheKey(kp CacheKeyParts) (CacheKey, error) {
+	data, err := json.Marshal(kp)
+
+	return CacheKey(data), err
+}
+
+// GetPresubmitsCached is like GetPresubmits, but uses a cache lookup to get the
+// *ProwYAML value (cache hit), instead of computing it from scratch (cache
+// miss). It also stores the entry into the cache if there is a cache miss.
+func (c *Config) GetPresubmitsCached(pc *ProwYAMLCache, gc git.ClientFactory, identifier string, baseSHAGetter RefGetter, headSHAGetters ...RefGetter) ([]Presubmit, error) {
```

Comment:
This isn't what I meant to describe. I don't think this really makes sense as a method of `Config`. It would be easier to use this if `ProwYAMLCache` was initialized with a reference to a config agent (or `config.Getter`) and a `git.ClientFactory`. Then we could just call `cache.GetPresubmits(identifer, baseSHAGetter, headSHAGetters...)` which is more concise.

This would also let us remove the need to pass a `git.ClientFactory` around separately. It would only be needed in the main function.
The `...MaybeCached` functions could also be dropped since we're either using the cache directly or not at all.

Reply:
done